win32/src/win32evtlog.i

```diff
@@ -511,6 +512,114 @@ static PyObject *PyEvtOpenChannelEnum(PyObject *self, PyObject *args, PyObject *
 }
 PyCFunction pfnPyEvtOpenChannelEnum = (PyCFunction) PyEvtOpenChannelEnum;
 
+
+// Helper function to convert a list of strings double zero terminated
+// into a Python list of strings.
+// e.g. hello\0world\0\0 -> [ "hello", "world" ]
+static PyObject* PyList_FromDoubleTerminatedWSTR(LPWSTR strings)
+{
+	PyObject* ret = PyList_New(0);
+	if (ret == NULL) {
+		return NULL;
+	}
+
+	WCHAR* cur = strings;
+	while (*cur) {
+		PyObject* keyword = PyWinObject_FromWCHAR(cur);
+		PyList_Append(ret, keyword);
+		Py_XDECREF(keyword);
+		cur += wcslen(cur) + 1;
+	}
+
+	return ret;
+}
+
+// Used internally to format event messages
+// Caller must free the returned value
+static PyObject *FormatMessageInternal(EVT_HANDLE metadata, EVT_HANDLE event, DWORD flags, DWORD resourceId)
+{
+	LPWSTR buf = NULL;
+	PyObject *ret = NULL;
+	DWORD allocated_size = 0;
+	DWORD returned_size = 0;
+	DWORD status = 0;
+	DWORD err = 0;
+
+	BOOL bsuccess = 0;
+	Py_BEGIN_ALLOW_THREADS
+	// Get the size of the buffer
+	bsuccess = EvtFormatMessage(metadata, event, resourceId, 0, NULL, flags, allocated_size, buf, &returned_size);
+	Py_END_ALLOW_THREADS
+
+	err = GetLastError();
+
+	// The above call should always return ERROR_INSUFFICIENT_BUFFER
+	if (!bsuccess && err != ERROR_INSUFFICIENT_BUFFER) {
+		PyWin_SetAPIError("EvtFormatMessage");
+	}
+
+	allocated_size = returned_size;
+	if (flags == EvtFormatMessageKeyword) {
+		allocated_size += 1; // +1 to double terminate the keyword list
+	}
+
+	allocated_size *= sizeof(WCHAR);
+	buf = (WCHAR *)malloc(allocated_size);
+	if (buf == NULL) {
+		PyErr_NoMemory();
+		return NULL;
+	}
+
+	Py_BEGIN_ALLOW_THREADS
+	bsuccess = EvtFormatMessage(metadata, event, resourceId, 0, NULL, flags, allocated_size, buf, &returned_size);
+	Py_END_ALLOW_THREADS
+
+	if (!bsuccess) {
+		char errorMessage[2048];
+		sprintf(errorMessage, "EvtFormatMessage: allocated %d, need buffer of size %d", allocated_size, returned_size);
+		PyWin_SetAPIError(errorMessage, err);
```

Comment:
I should have picked this before - but I'm not sure this is the right thing to do here. For one thing, `err` is now stale - and I'm really not sure the message is relevant -presumably the failure has nothing to do with the size of the buffer because we've correctly allocated what it mentions. So I think it's actually better to take the even simpler approach of just doing a `return PyWin_SetAPIError("EvtFormatMessage");` (which will do the right thing in terms of calling `GetLastError()` again.)